project-koku/koku
# [COST-2591] EBS to ORG-ID

## Jira Ticket

[COST-2591](https://issues.redhat.com/browse/COST-2591)

## Description

This change will require org-id for requests and identify customers and schema based on the org id of the customer.
- Customer objects are now filtered on org-id rather than account number
- Existing customers schemas do not change, they remain the same such as `acct10001`
- New tenants are created with the schema name of `org<org-id>` such as `org12345`
- HCS reports will include a new `org_id` column but will still include the `ebs_account_id` column
- New tenants S3 paths now use `org<org-id>` rather than account number in the path, such as:
```
koku-bucket/data/parquet/org2222222/AWS/source=30229a55-d138-4072-8d8c-1376a0e35946/year=2022/month=06
```
vs the current path of:
```
koku-bucket/data/parquet/10001/AWS/source=be82e2d3-e253-48ae-a372-f01ec4a93561/year=2022/month=06/
```

## Testing
This is a complex one to test, I will be providing ephemeral instructions so as to test the sources flow as well, if you need help with doing any of these steps, reach out:
1. Reserve an ephemeral environment with bonfire
2. Deploy the latest main to your namespace
3. Wait for everything to spin up
4. Port forward the sources service so you can send an api request: `oc port-forward service/sources-api-svc 3000:8000`
5. Send a source (using postman for example) to sources: `http://localhost:3000/api/sources/v3.1/bulk_create`
        a. You should include a few headers on this request, reach out for the relevant headers and examples if you need them
6. Once your source has processed, check your api customer table in the db in ephemeral: `SELECT * FROM api_customer;`, you should notice the org-id column is blank
7. Go ahead and update the api_customer and api_sources tables org-id column to match the org-id in your identity header from 5, you should know the account number as well: 
```
UPDATE api_customer SET org_id = '<org-id>' WHERE account_id = '<account-id>';
UPDATE api_sources SET org_id = '<org-id>' WHERE account_id = '<account-id>';
```
            a. in stage/prod these will be updated by a job
8. Port forward the api and make a few api requests and see some numbers: `oc port-forward service/koku-clowder-api 7500:8000` if you run into trouble hitting the api, you may need to adjust your header you are sending with the request.
9. Deploy this branch to your ephemeral environment, if you need some help with this, reach out
10. Send a new source with the same identity, you should see it get filled in nicely and your information in the database
11. Check a few api endpoints as well and see that your costs should have increased proportional to the new source
12. Send a new source with a new identity, the new schema will be based on the org-id and you can check the `api_customer` and `api_sources` tables and see your org-id is now listed in the column
13. Hit a few api endpoints with the new identity and see data for that customer

## Notes

...
